### DISCUSSION TITLE
Low Pitchrate and Rollrate

### DISCUSSION THREAD

**Discussion by beninaatwri** [JSBSim-Team/jsbsim#1208] [Support requests]
Hi,
I am working on integrating JSBSim with Unity. First, I created a Python script to capture input values from the HOTAS controller, which are then sent to the JSBSim script to process through the Flight Dynamics Model (FDM). The resulting calculations are transmitted to Unity to control the flight of a 3D aircraft. However, I feel that the pitch rate and roll rate are too low compared to flight gear. I would really appreciate it if someone could help me with this.  (Note: I have set FDM.Setdt to 0.01. I have attatched the python scripts along with this.)

[HotasControls.txt](https://github.com/user-attachments/files/18198364/HotasControls.txt)
[JSBSimPython.txt](https://github.com/user-attachments/files/18198366/JSBSimPython.txt)

**Comment by bcoconni** [comment]
It seems you're running one single simulation step of JSBSim for each iteration of your application loop however I guess you want real time behavior, so you have to make sure that JSBSim catches up with the elapsed real time. This is because JSBSim runs faster than real time so its simulated time lags behind the real world clock time.

For an example of how to implement this, check out the example below. It checks how much real time has elapsed since the last time JSBSim has been called and calls `fdm.run()` several times until JSBSim simulated time `fdm.get_sim_time()` equals the real world elapsed time `actual_elapsed_time`.

https://github.com/JSBSim-Team/jsbsim/blob/2bdf0d40ad2c39be0fa581f93cf1569883ef6077/python/JSBSim.py#L209-L214

**Reply by beninaatwri** [reply]
@bcoconni Thanks for the reply. You code really helped me to improve the simulation.

**Comment by bcoconni** [comment]
This support request has been moved to "Discussions" since it is neither a bug nor a feature request. The discussion will continue in this thread. Thank you.

**Comment by Warning39** [comment]
> It seems you're running one single simulation step of JSBSim for each iteration of your application loop however I guess you want real time behavior

Yep, taking a look the attached source code your code reading and processing the inputs is running in lock-step with JSBSim.

And in particular there is a big difference in the timesteps. So your code reading the inputs has a sleep of 0.1s (10Hz) after each network send of the inputs, and then the blocking TCP code on the other side runs a single timestep of JSBSim for which you've set the timestep to 0.01s (100Hz).

One option is to break the lock step. So have the code running JSBSim not block waiting for TCP input, make it a non-blocking read, and make use of the code above that @bcoconni posted to have the JSBSim simulation loop run at real time. Then whenever input does come in, e.g. at 10Hz via the non-blocking reads then feed the controls in then.

In this setup JSBSim will run at 100Hz real time and user inputs will be applied at 10Hz.

**Comment by beninaatwri** [comment]
Hi everyone,
Your replies have been incredibly helpful in improving the simulation. The synchronization between Unity and JSBSim is now excellent. However, I'm still experiencing some lag issues after running the simulation for a while. I'm unsure whether the lag is caused by JSBSim or Unity.

I've attached the updated code for your reference. Your assistance in resolving this issue would be greatly appreciated!

[Datareceiver_unity_updated.txt](https://github.com/user-attachments/files/18328914/Datareceiver_unity_updated.txt)
[JSBSimpython_updated.txt](https://github.com/user-attachments/files/18328915/JSBSimpython_updated.txt)
[Hotalcontrols_updated.txt](https://github.com/user-attachments/files/18328916/Hotalcontrols_updated.txt)
